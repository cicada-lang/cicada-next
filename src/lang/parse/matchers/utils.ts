import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
