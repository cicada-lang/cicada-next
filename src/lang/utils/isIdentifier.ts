const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
