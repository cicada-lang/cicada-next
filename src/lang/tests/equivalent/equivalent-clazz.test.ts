import { expect, test } from "vitest"
import { expectCodeToFail, runCode } from "../utils"

test("equivalent Clazz", async () => {
  const output = await runCode(`

function id(T: Type, x: T): T {
  return x
}

compute equivalent Type {
    class { A: Type, x: A }
  = class { A: Type, x: A }
  = class { A: Type, x: id(Type, A) }
}

`)

  expect(output).toMatchSnapshot()
})

test("equivalent Clazz -- Pair in class", async () => {
  const output = await runCode(`


compute equivalent Type {
    class { A: Type, B: Type, pair: Pair(A, B) }
  = class { A: Type, B: Type, pair: Pair(A, B) }
}

`)

  expect(output).toMatchSnapshot()
})

test("equivalent Clazz -- fail -- different property names", async () => {
  await expectCodeToFail(`

compute equivalent Type {
    class { A: Type, x: A }
  = class { B: Type, x: B }
}

`)
})

test("equivalent Clazz -- fail -- different property type", async () => {
  await expectCodeToFail(`

compute equivalent Type {
    class { A: Trivial }
  = class { A: String }
}

`)
})

test("equivalent Clazz -- fail -- different fulfilled property value", async () => {
  await expectCodeToFail(`

compute equivalent Type {
    class { A: String = "abc" }
  = class { A: String = "xyz" }
}

`)
})

test.todo(
  "equivalent Clazz -- fail -- missing fulfilled property value",
  async () => {
    await expectCodeToFail(`

compute equivalent Type {
    class { A: String = "abc" }
  = class { A: String }
}

`)

    await expectCodeToFail(`

compute equivalent Type {
    class { A: String }
  = class { A: String = "abc" }
}

`)
  },
)

test.todo("equivalent Clazz -- fail -- missing property type", async () => {
  await expectCodeToFail(`

compute equivalent Type {
    class { A: String, B: String }
  = class { A: String }
}

`)

  await expectCodeToFail(`

compute equivalent Type {
    class { A: String }
  = class { A: String, B: String }
}

`)
})
