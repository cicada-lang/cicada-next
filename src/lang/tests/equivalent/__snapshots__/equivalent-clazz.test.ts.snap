// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`equivalent Clazz > 
    "refl(
      implicit Type,
      implicit class {
        A: Type
        x [rename: x2]: A
      },
    ): Equal(
      Type,
      class {
        A: Type
        x [rename: x2]: A
      },
      class {
        A: Type
        x [rename: x2]: A
      },
    )"
   1`] = `
"refl(
  implicit Type,
  implicit class {
    A: Type
    x [rename: x2]: A
  },
): Equal(
  Type,
  class {
    A: Type
    x [rename: x2]: A
  },
  class {
    A: Type
    x [rename: x2]: A
  },
)"
`;

exports[`equivalent Clazz -- Pair in class > 
    "refl(
      implicit Type,
      implicit class {
        A: Type
        B: Type
        pair: Pair(A, B)
      },
    ): Equal(
      Type,
      class {
        A: Type
        B: Type
        pair: Pair(A, B)
      },
      class {
        A: Type
        B: Type
        pair: Pair(A, B)
      },
    )"
   1`] = `
"refl(
  implicit Type,
  implicit class {
    A: Type
    B: Type
    pair: Pair(A, B)
  },
): Equal(
  Type,
  class {
    A: Type
    B: Type
    pair: Pair(A, B)
  },
  class {
    A: Type
    B: Type
    pair: Pair(A, B)
  },
)"
`;

exports[`equivalent Clazz -- Pair in class: 
    "refl(
      implicit Type,
      implicit class {
        A: Type
        B: Type
        pair: Pair(A, B)
      },
    ): Equal(
      Type,
      class {
        A: Type
        B: Type
        pair: Pair(A, B)
      },
      class {
        A: Type
        B: Type
        pair: Pair(A, B)
      },
    )"
   1`] = `
"refl(
  implicit Type,
  implicit class {
    A: Type
    B: Type
    pair: Pair(A, B)
  },
): Equal(
  Type,
  class {
    A: Type
    B: Type
    pair: Pair(A, B)
  },
  class {
    A: Type
    B: Type
    pair: Pair(A, B)
  },
)"
`;

exports[`equivalent Clazz: 
    "refl(
      implicit Type,
      implicit class {
        A: Type
        x [rename: x2]: A
      },
    ): Equal(
      Type,
      class {
        A: Type
        x [rename: x2]: A
      },
      class {
        A: Type
        x [rename: x2]: A
      },
    )"
   1`] = `
"refl(
  implicit Type,
  implicit class {
    A: Type
    x [rename: x2]: A
  },
): Equal(
  Type,
  class {
    A: Type
    x [rename: x2]: A
  },
  class {
    A: Type
    x [rename: x2]: A
  },
)"
`;

exports[`equivalent Clazz 1`] = `
"refl(
  implicit Type,
  implicit class {
    A: Type
    x [rename: x2]: A
  },
): Equal(
  Type,
  class {
    A: Type
    x [rename: x2]: A
  },
  class {
    A: Type
    x [rename: x2]: A
  },
)"
`;

exports[`equivalent Clazz -- Pair in class 1`] = `
"refl(
  implicit Type,
  implicit class {
    A: Type
    B: Type
    pair: Pair(A, B)
  },
): Equal(
  Type,
  class {
    A: Type
    B: Type
    pair: Pair(A, B)
  },
  class {
    A: Type
    B: Type
    pair: Pair(A, B)
  },
)"
`;
